### PR TITLE
[SYCL] Allow accessor constructed with zero-size buffers

### PR DESCRIPTION
**Pull request by mmoadeli**
* Allow accessor constructed with zero-size buffers. [Clarify behaviour for range of zero](https://github.com/KhronosGroup/SYCL-Docs/pull/192)
* Remove existing error disallowing it.
* Add test

### CODE REVIEW

File: sycl/test-e2e/Basic/accessor/accessor.cpp [hunk 1]
```diff
@@ -837,17 +837,37 @@ int main() {
...
+      sycl::buffer<int, 1> b{0};
+
+      sycl::queue queue;
+      queue.submit([&](sycl::handler &cgh) {
+        auto B = b.template get_access<sycl::access::mode::read_write>(cgh);
+
+        cgh.single_task<class acc_with_zero_sized_buffer>([=]() {
+          if (!B.empty())
+            B[0] = 1;
+        });
```
Comment (steffenlarsen) [inline sycl/test-e2e/Basic/accessor/accessor.cpp:864]: Discussed offline. This should not throw.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 2]
```diff
@@ -3038,7 +3038,10 @@ pi_result cuda_piextKernelSetArgMemObj(pi_kernel kernel, pi_uint32 arg_index,
                                        const pi_mem *arg_value) {
 
   assert(kernel != nullptr);
-  assert(arg_value != nullptr);
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:3041]: Could we keep this check? You are dereferencing it below, so it still seems valid. Same goes for the HIP changes.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 3]
```diff
@@ -3038,7 +3038,10 @@ pi_result cuda_piextKernelSetArgMemObj(pi_kernel kernel, pi_uint32 arg_index,
...
+  if (*arg_value == nullptr) {
+    kernel->set_kernel_arg(arg_index, 0, nullptr);
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:3042]: Could you please add a comment to this addition to specify why it is needed. Same for the HIP changes.

File: sycl/source/detail/scheduler/commands.cpp [hunk 4]
```diff
@@ -2196,6 +2196,9 @@ static pi_result SetKernelParamsAndLaunch(
 
       RT::PiMem MemArg = (RT::PiMem)getMemAllocationFunc(Req);
       if (Queue->getDeviceImplPtr()->getBackend() == backend::opencl) {
+        if (!MemArg)
+          MemArg = RT::PiMem();
```
Comment (steffenlarsen) [inline sycl/source/detail/scheduler/commands.cpp:2200]: Please add a comment here as to why it is needed.

File: sycl/test-e2e/Basic/accessor/accessor.cpp (hunk 1 above)
Comment (mmoadeli) [reply sycl/test-e2e/Basic/accessor/accessor.cpp]: Thanks @steffenlarsen , the test are changed as advised.

File: sycl/plugins/cuda/pi_cuda.cpp (hunk 2 above)
Comment (mmoadeli) [reply sycl/plugins/cuda/pi_cuda.cpp:3041]: thansk @steffenlarsen , done

File: sycl/plugins/cuda/pi_cuda.cpp (hunk 3 above)
Comment (mmoadeli) [reply sycl/plugins/cuda/pi_cuda.cpp:3046]: thansk @steffenlarsen , done

File: sycl/source/detail/scheduler/commands.cpp (hunk 4 above)
Comment (mmoadeli) [reply sycl/source/detail/scheduler/commands.cpp:2204]: thansk @steffenlarsen , done.

File: sycl/test-e2e/Basic/accessor/accessor.cpp [hunk 5]
```diff
@@ -1435,5 +1407,38 @@ int main() {
...
+          .submit([&](sycl::handler &cgh) {
+            auto B =
+                Buf2.template get_access<sycl::access::mode::read_write>(cgh);
+
+            cgh.single_task<class acc_with_non_zero_sized_buffer>([=]() {
+              for (size_t I = 0; I < B.size(); ++I)
+                B[I] = 1;
+            });
+          })
+          .wait();
```
Comment (steffenlarsen) [inline sycl/test-e2e/Basic/accessor/accessor.cpp:1439]: Just to simplify it a little and make it easier to maintain, could we maybe do something like:

```suggestion
      for (auto &IBuf : {Buf, Buf2}) {
        queue
            .submit([&](sycl::handler &cgh) {
              auto B =
                  IBuf.template get_access<sycl::access::mode::read_write>(cgh);

              cgh.single_task<class fill_with_potentially_zero_size>([=]() {
                for (size_t I = 0; I < B.size(); ++I)
                  B[I] = 1;
              });
            })
            .wait();
      }
```

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 6]
```diff
@@ -3040,6 +3040,13 @@ pi_result cuda_piextKernelSetArgMemObj(pi_kernel kernel, pi_uint32 arg_index,
   assert(kernel != nullptr);
   assert(arg_value != nullptr);
 
+  // below sets kernel arg when zero-sized buffers are handled.
+  // in such case the corresponding memory is null.
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:3044]: Nit:
```suggestion
  // Below sets kernel arg when zero-sized buffers are handled.
  // In such case the corresponding memory is null.
```

File: sycl/plugins/hip/pi_hip.cpp [hunk 7]
```diff
@@ -2948,6 +2948,13 @@ pi_result hip_piextKernelSetArgMemObj(pi_kernel kernel, pi_uint32 arg_index,
   assert(kernel != nullptr);
   assert(arg_value != nullptr);
 
+  // below sets kernel arg when zero-sized buffers are handled.
+  // in such case the corresponding memory is null.
```
Comment (steffenlarsen) [inline sycl/plugins/hip/pi_hip.cpp:2952]:
```suggestion
  // Below sets kernel arg when zero-sized buffers are handled.
  // In such case the corresponding memory is null.
```

**Review by steffenlarsen — APPROVED** [review]
Small nits, otherwise LGTM!

**Review by jchlanda — APPROVED** [review]

**Pull request merged**